base/src/main/java/one/microstream/collections/lazy/LazyArrayList.java

```diff
@@ -0,0 +1,1367 @@
+package one.microstream.collections.lazy;
+
+/*-
+ * #%L
+ * microstream-base
+ * %%
+ * Copyright (C) 2019 - 2022 MicroStream Software
+ * %%
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is
+ * available at https://www.gnu.org/software/classpath/license.html.
+ * 
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ * #L%
+ */
+
+import static one.microstream.collections.AbstractExtendedCollection.validateIndex;
+import static one.microstream.math.XMath.positive;
+
+import java.lang.reflect.Array;
+import java.util.AbstractList;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.ConcurrentModificationException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.RandomAccess;
+import java.util.Spliterator;
+import java.util.Stack;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+
+import one.microstream.branching.ThrowBreak;
+import one.microstream.reference.Lazy;
+import one.microstream.reference.LazyClearObserver;
+import one.microstream.reference.ObservedLazyReference;
+
+
+/**
+ * {@link LazyList} implementation according to the specifications of {@link ArrayList}.
+ * <p>
+ * Lazy-loaded segments are used internally to achieve the partial loading.
+ * The maximum size of these segments can be specified within the constructor.
+ * <br>
+ * This implementation requires an active microstream storage with specialized
+ * type handlers. Without those handles a correct behavior is not guaranteed.
+ * The required handlers are:
+ * BinaryHandlerLazyArrayList
+ * BinaryHandlerLazyObservable
+ * <br>
+ * This list tries to unload segments depending on the provided {@link LazySegmentUnloader}.
+ * By default, it will use the {@link LazySegmentUnloader.Default} that keeps two segments loaded.
+ *
+ * @param <E> the type of elements in this collection
+ */
+public class LazyArrayList<E> extends AbstractList<E> implements LazyList<E>, RandomAccess
```

Comment:
Making all lazy implementations 'final' is probably a good idea.

Reply:
agreed, done